src/uri.rs

```diff
@@ -207,13 +250,19 @@ impl FromStr for Uri {
     type Err = Error;
 
     fn from_str(s: &str) -> Result<Uri, Error> {
-        Uri::new(s)
+        Uri::new(InternalUri::from(s))
     }
 }
 
 impl From<Url> for Uri {
     fn from(url: Url) -> Uri {
-        Uri::new(url.as_str()).expect("Uri::From<Url> failed")
+        Uri::new(InternalUri::from(url.as_str())).expect("Uri::From<Url> failed")
+    }
+}
+
+impl From<MemStr> for Uri {
+    fn from(url: MemStr) -> Uri {
+        Uri::new(InternalUri::from(url.as_str())).expect("Uri::From<MemStr> failed")
```

Comment:
`From` implementations should be for conversions that cannot fail. As mentioned earlier, we receive a bunch of bytes from the socket, and after some `httparse` calls, we know the URI is a string, but we don't know if it is a valid `Uri`. It needs to go through `Uri::new` to parse it, and that could fail. It's better to not panic.

I've updated master with a test case in the parse.rs module that will trigger this case, so if you rebase you should see the panic (that should be an `Err` instead).